File: macos/Sources/UpdateChecker.swift
import AppKit
import Foundation
import Combine

final class UpdateChecker: ObservableObject {
    static let releasesPage = URL(string: "https://github.com/BadryansahBangsawan/MonitorSuhu/releases/latest")!
    private static let api = URL(string: "https://api.github.com/repos/BadryansahBangsawan/MonitorSuhu/releases?per_page=30")!

    let currentVersion: String

    @Published private(set) var latestVersion: String?
    @Published private(set) var latestURL: URL?
    @Published private(set) var latestAsset: ReleaseAsset?
    @Published private(set) var hasUpdate = false
    @Published private(set) var checking = false
    @Published private(set) var installing = false
    @Published private(set) var progress: Double = 0
    @Published private(set) var lastError: String?

    init(currentVersion: String = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0") {
        self.currentVersion = currentVersion
    }

    func check(userInitiated: Bool = false, completion: ((Bool) -> Void)? = nil) {
        guard !checking, !installing else { return }
        checking = true
        lastError = nil

        var request = URLRequest(url: Self.api, timeoutInterval: 12)
        request.setValue("MonitorSuhu/\(currentVersion)", forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.checking = false
                if let error {
                    self.lastError = error.localizedDescription
                    completion?(false)
                    if userInitiated { Self.alert("Could not check for updates.", error.localizedDescription) }
                    return
                }
                let status = (response as? HTTPURLResponse)?.statusCode ?? 0
                guard status == 200, let data,
                      let json = try? JSONSerialization.jsonObject(with: data)
                else {
                    self.lastError = "GitHub returned no release."
                    completion?(false)
                    if userInitiated { Self.alert("Could not check for updates.", "GitHub returned no release.") }
                    return
                }
                let blobs: [[String: Any]]
                if let many = json as? [[String: Any]] {
                    blobs = many
                } else if let one = json as? [String: Any] {
                    blobs = [one]
                } else {
                    blobs = []
                }
                let releases = blobs.compactMap(Self.parseRelease)
                guard let hit = ReleaseAssets.latest(in: releases, platform: "macos") else {
                    self.latestVersion = nil
                    self.latestURL = Self.releasesPage
                    self.latestAsset = nil
                    self.hasUpdate = false
                    self.lastError = "No macOS build on GitHub Releases."
                    completion?(true)
                    if userInitiated {
                        Self.alert("No macOS update.", "GitHub latest has no macOS disk image. Windows or Linux-only releases are ignored here.")
                    }
                    return
                }
                self.latestVersion = hit.tag
                self.latestURL = hit.url ?? Self.releasesPage
                self.latestAsset = hit.asset
                self.hasUpdate = Versioning.isNewer(hit.tag, than: self.currentVersion)
                completion?(true)
                if userInitiated {
                    if self.hasUpdate {
                        self.promptInstall(latest: hit.tag)
                    } else {
                        Self.alert("You’re up to date.", "MonitorSuhu \(self.currentVersion) is the latest macOS release.")
                    }
                }
            }
        }.resume()
    }

    func apply() {
        guard !installing else { return }
        guard let asset = latestAsset else {
            openDownloadPage()
            return
        }
        installing = true
        progress = 0
        lastError = nil
        let agent = "MonitorSuhu/\(currentVersion)"
        Task.detached { [weak self] in
            do {
                let destDir = FileManager.default.temporaryDirectory
                    .appendingPathComponent("MonitorSuhu-update", isDirectory: true)
                try FileManager.default.createDirectory(at: destDir, withIntermediateDirectories: true)
                let file = destDir.appendingPathComponent(asset.name)
                try await Self.download(asset, to: file, userAgent: agent) { value in
                    Task { @MainActor [weak self] in
                        self?.progress = value * 0.85
                    }
                }
                await MainActor.run { [weak self] in
                    self?.progress = 1
                }
                try UpdateInstaller.apply(dmg: file)
            } catch {
                let message = error.localizedDescription
                await MainActor.run { [weak self] in
                    self?.installing = false
                    self?.lastError = message
                    Self.alert("Could not install update.", message)
                }
            }
        }
    }

    func openDownloadPage() {
        NSWorkspace.shared.open(latestURL ?? Self.releasesPage)
    }

    static func isNewer(_ latest: String, than current: String) -> Bool {
        Versioning.isNewer(latest, than: current)
    }

    private func promptInstall(latest: String) {
        let alert = NSAlert()
        alert.messageText = "MonitorSuhu \(latest) is available"
        alert.informativeText = latestAsset == nil
            ? "Download it from GitHub Releases, then replace the app in Applications."
            : "Download and replace the app in place. Your settings stay where they are."
        alert.alertStyle = .informational
        alert.addButton(withTitle: latestAsset == nil ? "Download" : "Install")
        alert.addButton(withTitle: "Later")
        if alert.runModal() == .alertFirstButtonReturn {
            if latestAsset == nil {
                openDownloadPage()
            } else {
                apply()
            }
        }
    }

    private static func parseRelease(_ body: [String: Any]) -> GitHubRelease? {
        guard let tag = body["tag_name"] as? String else { return nil }
        let page = (body["html_url"] as? String).flatMap(URL.init(string:))
        let draft = body["draft"] as? Bool ?? false
        let pre = body["prerelease"] as? Bool ?? false
        let raw = body["assets"] as? [[String: Any]] ?? []
        let assets: [ReleaseAsset] = raw.compactMap { item in
            guard let name = item["name"] as? String,
                  let href = item["browser_download_url"] as? String,
                  let url = URL(string: href)
            else { return nil }
            let size = (item["size"] as? NSNumber)?.int64Value ?? 0
            return ReleaseAsset(name: name, url: url, size: size)
        }
        return GitHubRelease(tag: tag, htmlURL: page, assets: assets, draft: draft, prerelease: pre)
    }

    private static func download(
        _ asset: ReleaseAsset,
        to dest: URL,
        userAgent: String,
        progress: @escaping (Double) -> Void
    ) async throws {
        guard ReleaseAssets.isTrusted(asset.url, name: asset.name, platform: "macos") else {
            throw NSError(domain: "MonitorSuhu", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Update file is not from GitHub Releases."
            ])
        }
        if asset.size > ReleaseAssets.maxBytes {
            throw NSError(domain: "MonitorSuhu", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Update is larger than expected."
            ])
        }

        var request = URLRequest(url: asset.url, timeoutInterval: 300)
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        let (temp, response) = try await URLSession.shared.download(for: request)
        let http = response as? HTTPURLResponse
        guard http?.statusCode == 200 else {
            throw NSError(domain: "MonitorSuhu", code: http?.statusCode ?? 0, userInfo: [
                NSLocalizedDescriptionKey: "GitHub returned no file."
            ])
        }
        let length = http?.expectedContentLength ?? asset.size
        if length > ReleaseAssets.maxBytes {
            throw NSError(domain: "MonitorSuhu", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Update is larger than expected."
            ])
        }
        let written = (try? FileManager.default.attributesOfItem(atPath: temp.path)[.size] as? NSNumber)?.int64Value ?? 0
        if written > ReleaseAssets.maxBytes {
            throw NSError(domain: "MonitorSuhu", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Update is larger than expected."
            ])
        }
        if asset.size > 0 && written != asset.size {
            throw NSError(domain: "MonitorSuhu", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Download size did not match GitHub."
            ])
        }
        progress(1)
        if FileManager.default.fileExists(atPath: dest.path) {
            try FileManager.default.removeItem(at: dest)
        }
        try FileManager.default.moveItem(at: temp, to: dest)
    }

    private static func alert(_ title: String, _ message: String) {
        let alert = NSAlert()
        alert.messageText = title
        alert.informativeText = message
        alert.alertStyle = .informational
        alert.addButton(withTitle: "OK")
        alert.runModal()
    }
}

File: windows/src/MonitorSuhu.App/Services/UpdateChecker.cs
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using CommunityToolkit.Mvvm.ComponentModel;
using MonitorSuhu.Core;

namespace MonitorSuhu.App.Services;

public sealed partial class UpdateChecker : ObservableObject
{
    public const string ReleasesPage = "https://github.com/BadryansahBangsawan/MonitorSuhu/releases/latest";
    private const string Api = "https://api.github.com/repos/BadryansahBangsawan/MonitorSuhu/releases?per_page=30";

    public string CurrentVersion { get; }

    [ObservableProperty] private string? latestVersion;
    [ObservableProperty] private string? latestUrl;
    [ObservableProperty] private ReleaseAsset? latestAsset;
    [ObservableProperty] private bool hasUpdate;
    [ObservableProperty] private bool checking;
    [ObservableProperty] private bool installing;
    [ObservableProperty] private double progress;
    [ObservableProperty] private string? lastError;

    public string UpdateMessage =>
        Installing
            ? Progress < 1
                ? $"Downloading MonitorSuhu {LatestVersion}…"
                : $"Installing MonitorSuhu {LatestVersion}…"
            : HasUpdate && !string.IsNullOrEmpty(LatestVersion)
                ? $"MonitorSuhu {LatestVersion} is ready."
                : $"MonitorSuhu {CurrentVersion}";

    public string UpdateHint =>
        Installing
            ? "The app will close to finish. Open MonitorSuhu again from the Start menu if it does not restart."
            : HasUpdate
                ? "Install from here. Settings stay in AppData. The app will close — reopen it if it does not come back."
                : "";

    partial void OnInstallingChanged(bool value) => NotifyCopy();
    partial void OnProgressChanged(double value) => NotifyCopy();
    partial void OnHasUpdateChanged(bool value) => NotifyCopy();

    private void NotifyCopy()
    {
        OnPropertyChanged(nameof(UpdateMessage));
        OnPropertyChanged(nameof(UpdateHint));
    }

    public UpdateChecker(string? currentVersion = null)
    {
        var v = currentVersion
            ?? typeof(UpdateChecker).Assembly.GetName().Version?.ToString()
            ?? "0";
        CurrentVersion = Versioning.Trim(v);
    }

    public async Task CheckAsync(bool userInitiated = false)
    {
        if (Checking || Installing) return;
        Checking = true;
        LastError = null;
        try
        {
            using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(12) };
            client.DefaultRequestHeaders.UserAgent.ParseAdd($"MonitorSuhu/{CurrentVersion}");
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
            using var response = await client.GetAsync(Api);
            var json = await response.Content.ReadAsStringAsync();
            if (!response.IsSuccessStatusCode)
            {
                LastError = $"GitHub HTTP {(int)response.StatusCode}";
                if (userInitiated)
                    Show("Could not check for updates.", LastError);
                return;
            }

            using var doc = JsonDocument.Parse(json);
            var hit = ReleaseAssets.LatestFor(ParseReleases(doc.RootElement), "windows");
            if (hit is null)
            {
                LatestVersion = null;
                LatestUrl = ReleasesPage;
                LatestAsset = null;
                HasUpdate = false;
                LastError = "No Windows build on GitHub Releases.";
                OnPropertyChanged(nameof(UpdateMessage));
                OnPropertyChanged(nameof(UpdateHint));
                if (userInitiated)
                    Show("No Windows update.", "GitHub latest has no Windows installer. macOS or Linux-only releases are ignored here.");
                return;
            }

            LatestVersion = hit.Value.Tag;
            LatestUrl = string.IsNullOrWhiteSpace(hit.Value.HtmlUrl) ? ReleasesPage : hit.Value.HtmlUrl;
            LatestAsset = hit.Value.Asset;
            HasUpdate = Versioning.IsNewer(hit.Value.Tag, CurrentVersion);
            OnPropertyChanged(nameof(UpdateMessage));
            OnPropertyChanged(nameof(UpdateHint));

            if (userInitiated)
            {
                if (HasUpdate)
                    PromptInstall(hit.Value.Tag);
                else
                    Show("You're up to date.", $"MonitorSuhu {CurrentVersion} is the latest Windows release.");
            }
        }
        catch (Exception ex)
        {
            LastError = ex.Message;
            if (userInitiated)
                Show("Could not check for updates.", ex.Message);
        }
        finally
        {
            Checking = false;
        }
    }

    public async Task ApplyAsync()
    {
        if (Installing) return;
        if (LatestAsset is null)
        {
            OpenDownloadPage();
            return;
        }

        Installing = true;
        Progress = 0;
        LastError = null;
        try
        {
            var dir = System.IO.Path.Combine(System.IO.Path.GetTempPath(), "MonitorSuhu-update");
            System.IO.Directory.CreateDirectory(dir);
            var dest = System.IO.Path.Combine(dir, LatestAsset.Name);
            var progress = new Progress<double>(p => Progress = p);
            await UpdateDownload.ToFileAsync(
                LatestAsset.Url,
                dest,
                $"MonitorSuhu/{CurrentVersion}",
                LatestAsset.Size,
                "windows",
                progress);
            Progress = 1;
            await Task.Delay(1800);
            UpdateInstaller.LaunchWindows(dest);
            System.Windows.Application.Current?.Shutdown();
        }
        catch (Exception ex)
        {
            LastError = ex.Message;
            Installing = false;
            Show("Could not install update.", ex.Message);
        }
    }

    public void OpenDownloadPage()
    {
        var url = string.IsNullOrWhiteSpace(LatestUrl) ? ReleasesPage : LatestUrl;
        Process.Start(new ProcessStartInfo(url) { UseShellExecute = true });
    }

    public static bool IsNewer(string latest, string current) =>
        Versioning.IsNewer(latest, current);

    private void PromptInstall(string tag)
    {
        var canInstall = LatestAsset is not null;
        var message = canInstall
            ? "Download and install it now. Settings stay in AppData. The app will close — open MonitorSuhu again from the Start menu if it does not restart."
            : "Download it from GitHub Releases, run the installer, then open MonitorSuhu again.";
        if (!canInstall)
        {
            Show($"MonitorSuhu {tag} is available", message, open: LatestUrl ?? ReleasesPage);
            return;
        }

        var result = System.Windows.MessageBox.Show(
            message + "\n\nInstall now?",
            $"MonitorSuhu {tag} is available",
            System.Windows.MessageBoxButton.YesNo);
        if (result == System.Windows.MessageBoxResult.Yes)
            _ = ApplyAsync();
    }

    private static List<GitHubRelease> ParseReleases(JsonElement root)
    {
        var list = new List<GitHubRelease>();
        if (root.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in root.EnumerateArray())
            {
                if (ParseRelease(item) is { } release)
                    list.Add(release);
            }
        }
        else if (ParseRelease(root) is { } one)
        {
            list.Add(one);
        }
        return list;
    }

    private static GitHubRelease? ParseRelease(JsonElement root)
    {
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (!root.TryGetProperty("tag_name", out var tagEl)) return null;
        var tag = tagEl.GetString() ?? "";
        if (string.IsNullOrEmpty(tag)) return null;
        string? page = null;
        if (root.TryGetProperty("html_url", out var html))
            page = html.GetString();
        var draft = root.TryGetProperty("draft", out var d) && d.ValueKind == JsonValueKind.True;
        var pre = root.TryGetProperty("prerelease", out var p) && p.ValueKind == JsonValueKind.True;
        return new GitHubRelease(tag, page, ParseAssets(root), draft, pre);
    }

    private static List<ReleaseAsset> ParseAssets(JsonElement root)
    {
        if (!root.TryGetProperty("assets", out var assets) || assets.ValueKind != JsonValueKind.Array)
            return [];
        var list = new List<ReleaseAsset>();
        foreach (var item in assets.EnumerateArray())
        {
            var name = item.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
            var url = item.TryGetProperty("browser_download_url", out var u) ? u.GetString() ?? "" : "";
            var size = item.TryGetProperty("size", out var s) && s.TryGetInt64(out var bytes) ? bytes : 0;
            if (!string.IsNullOrEmpty(name) && !string.IsNullOrEmpty(url))
                list.Add(new ReleaseAsset(name, url, size));
        }
        return list;
    }

    private static void Show(string title, string message, string? open = null)
    {
        var app = System.Windows.Application.Current;
        if (app?.Dispatcher is { } dispatcher && !dispatcher.CheckAccess())
        {
            dispatcher.Invoke(() => Show(title, message, open));
            return;
        }

        if (open is null)
        {
            System.Windows.MessageBox.Show(message, title, System.Windows.MessageBoxButton.OK);
            return;
        }

        var result = System.Windows.MessageBox.Show(
            message + "\n\nOpen the download page?",
            title,
            System.Windows.MessageBoxButton.YesNo);
        if (result == System.Windows.MessageBoxResult.Yes)
            Process.Start(new ProcessStartInfo(open) { UseShellExecute = true });
    }
}
